include/core/mir/mutex.h

```diff
@@ -87,13 +102,24 @@ class Mutex
      *          While code has access to the MutexGuard it is guaranteed to have exclusive
      *          access to the contained data.
      */
-    MutexGuard<Guarded> lock()
+    auto lock() -> MutexGuard<Guarded>
     {
         return MutexGuard<Guarded>{std::unique_lock{mutex}, value};
     }
 
+    /**
+     * Lock the mutex and return an accessor for the protected data.
+     *
+     * \return A smart-pointer-esque accessor for the contained data.
+     *         While code has access to the MutexGuard it is guaranteed to have exclusive
+     *         access to the contained data.
+     */
+    auto lock() const -> MutexGuard<Guarded const>
```

Comment:
Hm. I suspect that this will be a compile error if someone asks for `Mutex<Foo const>`, right? Do we care about doing the necessary template shenanigans to make that work? `Mutex<Foo const>` doesn't seem like a useful thing to ask for.

Reply:
I agree that `Mutex<Foo const>` "doesn't seem like a useful thing to ask for", but would there be a compile error?